### PR TITLE
ENH: Initial configuration to introduce pre-commit

### PR DESCRIPTION
**Pull request by hjmjohnson**
Update SetupForDevelopment to setup pre-commit

Using a temporary repo to hold Kitware hooks with a configuration for
pre-commit. Replaces the system to clone the hook repo/branch into the
.git/hooks directory, which just using pre-commit.

This PR mirrors work from SimpleITK performed by
Bradley Lowekamp <blowekamp@mail.nih.gov>

The pre-commit python virtual environment can be manually
loaded with:  source .git/hooks/venv/bin/activate

which would allow using the pre-commit command prior to
doing an actual commit.

i.e.
```bash
source .git/hooks/venv/bin/activate
pre-commit run -a
```

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)

Related to:  https://github.com/InsightSoftwareConsortium/ITK/issues/4238
Related to:  https://github.com/InsightSoftwareConsortium/ITK/issues/4565
Related to:  https://github.com/InsightSoftwareConsortium/ITK/issues/4244

### CODE REVIEW

**Comment by blowekamp** [comment]
I made an upstream pull request for Kitware's GitSetup to support precommit to avoid the code duplication:
https://gitlab.kitware.com/utils/gitsetup/-/issues/18#note_1499602
But it was rejected due to it being more python specific and not as generic and efficient as the shell system infrastructure.

Also in #4651 I started to use clang-format into pre-commit.

Sometimes with big rebases and commit the large number of recommit checks is slower, so perhaps there can be too many checks.  However I have found the precommit file easier to maintain.

**Review by jhlegarreta — COMMENTED** [review]
Thanks for doing this Hans.

Have not tried locally.

A few comments:
- Should this file be deleted?
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/GitSetup/setup-hooks

- Should these lines be removed:
https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/GitSetup/config#L1-L2

- Not sure if I understand the passage "Using a temporary repo to hold Kitware hooks with a configuration for pre-commit.". Which is the temporary repository?

**Comment by blowekamp** [comment]
> > I made an upstream pull request for Kitware's GitSetup to support precommit to avoid the code duplication: https://gitlab.kitware.com/utils/gitsetup/-/issues/18#note_1499602 But it was rejected due to it being more python specific and not as generic and efficient as the shell system infrastructure.
> > Also in #4651 I started to use clang-format into pre-commit.
> > Sometimes with big rebases and commit the large number of recommit checks is slower, so perhaps there can be too many checks. However I have found the precommit file easier to maintain.
> 
> Pre-commit is tremendously easier to maintain. I don't think the performance concerns are that disruptive.

Yes, I agree it's likely worth the switch. Just need to consider ways to reduce code duplication etc.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+   - id: black
+     args: ['--diff', '-v'] # if run without arguments, will fail and will format the files
+- repo: local
+  hooks:
+    - id: local prepare-commit-msg
+      name: 'local prepare-commit-msg'
+      entry: 'Utilities/Hooks/prepare-commit-msg'
+      language: system
+      stages: [prepare-commit-msg]
+    - id: local pre-commit
```
Comment (blowekamp) [inline .pre-commit-config.yaml:50]: There is not "check-setup-for-development" hook here. It this intentional?

Comment (hjmjohnson) [reply .pre-commit-config.yaml:50]: Just missed it.  Thank you for identifying.

**Comment by blowekamp** [comment]
What was proposed in: https://gitlab.kitware.com/utils/gitsetup/-/issues/18 was to add a .pre-commit-hooks.yaml file to the upstream GitSetup repository. This would allow project's .pre-commit-config.yaml files to refer to the repository, which would remove the need for the local scripts, and simplify the local config file a little. We could for the GitSetup repo into our org and add the required file.

Should we keep those files locally or use a fork?

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+     exclude: "\\.(sha|sha512|svg|vtk|vtp)$|\\/ThirdParty\\/|\\/Data\\/"
+   - id: name-tests-test
+     args: ['--pytest-test-first']
+   - id: trailing-whitespace
+     exclude: "\\.(sha|sha512|svg|vtk|vtp)$|\\/ThirdParty\\/|\\/Data\\/"
+- repo: https://github.com/psf/black
+  rev: 24.2.0
+  hooks: # check conformance to black formatting
+   - id: black
+     name: black
```
Comment (blowekamp) [inline .pre-commit-config.yaml:42]: Why are these parameters being duplicated? They should be coming from this file in the repo:
https://github.com/psf/black/blob/main/.pre-commit-hooks.yaml

Comment (hjmjohnson) [reply .pre-commit-config.yaml:42]: I was just being explicit here to document what the flags are.  

The source you list is where I copied them from.

They could be removed, and the defaults could be used.

**Comment by hjmjohnson** [comment]
> Should we keep those files locally or use a fork?

Let's keep the files locally for now and optimize them across projects later. Adding external dependencies and external reliance on others to make changes could complicate the initial switch to pre-commit.

**Comment by blowekamp** [comment]
> Let's keep the files locally for now and optimize them across projects later. Adding external dependencies and external reliance on others to make changes could complicate the initial switch to pre-commit.

For these particular file they already exist in the "hooks" branch: https://github.com/InsightSoftwareConsortium/ITK/tree/hooks We may be able to continue to use them from there, and just add that hooks file to the branch.

**Comment by hjmjohnson** [comment]
@dzenanz @blowekamp This should be ready for final reviews now.  I have it working on my Mac, and older testing that I think is the same, worked on windows and linux.

File: CMake/ITKModuleMacros.cmake [hunk 3]
```diff
@@ -6,7 +6,6 @@ include(${_ITKModuleMacros_DIR}/ITKModuleAPI.cmake)
 include(${_ITKModuleMacros_DIR}/ITKModuleDoxygen.cmake)
 include(${_ITKModuleMacros_DIR}/ITKModuleHeaderTest.cmake)
 include(${_ITKModuleMacros_DIR}/ITKModuleKWStyleTest.cmake)
-include(${_ITKModuleMacros_DIR}/ITKModuleClangFormat.cmake)
```
Comment (dzenanz) [inline CMake/ITKModuleMacros.cmake:9]: @N-Dekker Ahh, with this commit, we will not be required to "know" about existence of `clang-format`. Except for use in IDEs, such as Visual Studio 😄

**Review by dzenanz — APPROVED** [review]

**Review by blowekamp — APPROVED** [review]

**Comment by blowekamp** [comment]
I just encountered an issue with updating ThirdParty libraries and the precommit setup:
```
++ git commit --no-gpg-sign -n '--author=OpenJPEG Upstream <kwrobot@kitware.com>' '--date=2024-12-09 17:31:23 +0100' -F -
No .pre-commit-config.yaml file was found
- To temporarily silence this, run `PRE_COMMIT_ALLOW_NO_CONFIG=1 git ...`
- To permanently silence this, install pre-commit with the --allow-missing-config option
- To uninstall pre-commit run `pre-commit uninstall`
```

**Pull request merged**